### ISSUE TITLE
deserializing JSON to object with enum causes BadFormatError

### ISSUE THREAD

**Issue by hokiegeek2** [chapel-lang/chapel#18335]
### Summary of Problem
Serializing an object with an enum works fine w/ no errors. Deserializing that JSON throws a BadFormatError, although the object is actually created correctly. This is not a blocking issue b/c the logic actually works, but deserializing requires a try...catch block.

Note: this deserialization error does not occur with Records, just Classes.

### Steps to Reproduce

```
module JsonFun {
    use IO;
    use Reflection;

    proc main() {
        enum CounterField{NUM_REQUESTS};

        class Metric {
            var metricName: string;
        }

        class CounterMetric : Metric{
            var value: int;
            var field: CounterField;
        }

        try! writeln('%jt'.format(new CounterMetric('requests_export',1,CounterField.NUM_REQUESTS)));

        var cm = new CounterMetric();
        try! deserialize(cm,'{"metricName":"requests_export", "value":1, "field": "CounterField.NUM_REQUESTS"}');
        try! writeln(cm);
    }

    proc deserialize(ref msg, request) throws {
        var newmem = openmem();
        newmem.writer().write(request);
        var nreader = newmem.reader();
        try {
            nreader.readf("%jt", msg);
        } catch bfe : BadFormatError {
            try! writeln(bfe);
        }
    }
}
```
Executing the code results in the following:

![image](https://user-images.githubusercontent.com/10785153/131911870-8895ac9a-0e76-4d4d-b3fa-9c2a3b570c53.png)

**Compile command:**
chpl JsonFun.chpl

**Execution command:**
<!-- e.g. `./foo -nl 4`
./JsonFun -nl 1

**Associated Future Test(s):**
No
### Configuration Information

- Output of `chpl --version`:
```
chpl version 1.24.1
Copyright 2020-2021 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)
```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: clang
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: smp *
  CHPL_GASNET_SEGMENT: fast
CHPL_TASKS: qthreads
CHPL_LAUNCHER: smp
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.3 (clang-1103.0.32.59)
Target: x86_64-apple-darwin19.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by cassella** [comment]
I don't think it's actually working -- you're just getting unfortunately lucky because the enum value you're deserializing has the value 0.  If I add a few new values to the enum and deserialize a string with one of those, it still reports `NUM_REQUESTS`, because the error is being thrown before filling in that field.


```chapel
        try! deserialize(cm,'{"metricName":"requests_export", "value":1, "field": "CounterField.OTHER2"}');
```

```
$ ./foo -nl 1
{"metricName":"requests_export", "value":1, "field":"OTHER2"}
BadFormatError: bad format: missing expected literal (while reading CounterField with path "unknown" offset 52)
{metricName = requests_export, value = 1, field = NUM_REQUESTS}
$ 
```



However, it works with no error and the correct value if I leave off the `CounterField.` prefix on the enum value in the string:

```chapel
        try! deserialize(cm,'{"metricName":"requests_export", "value":1, "field": "OTHER2"}');
```

```
$ ./foo -nl 1
{"metricName":"requests_export", "value":1, "field":"OTHER2"}
{metricName = requests_export, value = 1, field = OTHER2}
$ 
```

(Which is how the `writeln('%jt',)` formats it.)

**Comment by hokiegeek2** [comment]
@cassella Gotcha, the problem is that I want Chapel to deserialize the string field into an enum for the purposes of controlled vocab, which is why I want to pass in CounterField.NUM_REQUESTS.

**Comment by hokiegeek2** [comment]
wow, it gets worse:

```
   use IO;
    use Reflection;

    proc main() {
        enum CounterField{NUM_REQUESTS,NUM_ERRORS,NUM_NEW_CONNECTIONS};

        class Metric {
            var metricName: string;
        }

        class CounterMetric : Metric{
            var value: int;
            var field: CounterField;
        }

        try! writeln('%jt'.format(new CounterMetric('requests_export',1,CounterField.NUM_REQUESTS)));

        var cm = new CounterMetric();
        try! deserialize(cm,'{"metricName":"requests_export", "value":1, "field": "NUM_REQUESTS"}');
        try! writeln('Not a CounterMetric %s'.format(cm.field));
        try! writeln(cm.field == CounterField.NUM_REQUESTS);

        try! writeln('%jt'.format(new CounterMetric('requests_export',1,CounterField.NUM_NEW_CONNECTIONS)));
        cm = new CounterMetric();
        try! deserialize(cm,'{"metricName":"requests_export", "value":1, "field": "NUM_NEW_CONNECTIONS}');
        try! writeln(cm.field == CounterField.NUM_NEW_CONNECTIONS);

    }

    proc deserialize(ref msg, request) throws {
        var newmem = openmem();
        newmem.writer().write(request);
        var nreader = newmem.reader();
        try {
            writeln("INCOMING MSG: %s".format(request));
            nreader.readf("%jt", msg);
            writeln("READ MSG: %s".format(request));
        } catch bfe : BadFormatError {
            try! writeln(bfe);
        }
    }
}
```

![image](https://user-images.githubusercontent.com/10785153/132094364-25814918-f0f8-42bb-8b3a-296ba4deb978.png)

This shows that the deserialization to an enum fails even if CounterField is left off for my example that used the third enum value.

**Comment by cassella** [comment]
The bad format error in this case is probably about the missing `"` towards the end of the string `'{... "field": "NUM_NEW_CONNECTIONS}'` in the second call to `deserialize`.

**Comment by mppf** [comment]
hi @hokiegeek2 - I've merged PR #18356 to address this issue and I'm thinking that we should close this issue. Could you let me know if that makes sense to you? Thanks.

**Comment by hokiegeek2** [comment]
@mppf works for me! thanks a bunch for making this happen.
